src/mapstorage/XmlMapStorage.cpp

```diff
@@ -0,0 +1,462 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright (C) 2021 The MMapper Authors
+// Author: Massimiliano Ghilardi <massimiliano.ghilardi@gmail.com> (Cosmos)
+// Author: Nils Schimmelmann <nschimme@gmail.com> (Jahara)
+
+#include "XmlMapStorage.h"
+
+#include <cassert>
+#include <cstddef>
+#include <stdexcept>
+#include <QString>
+#include <QXmlStreamWriter>
+
+#include "../expandoracommon/coordinate.h"
+#include "../expandoracommon/exit.h"
+#include "../expandoracommon/room.h"
+#include "../global/roomid.h"
+#include "../global/utils.h"
+#include "../mapdata/DoorFlags.h"
+#include "../mapdata/ExitDirection.h"
+#include "../mapdata/ExitFlags.h"
+#include "../mapdata/enums.h"
+#include "../mapdata/mapdata.h"
+#include "../mapdata/mmapper2room.h"
+#include "abstractmapstorage.h"
+#include "basemapsavefilter.h"
+#include "progresscounter.h"
+#include "roomsaver.h"
+
+XmlMapStorage::XmlMapStorage(MapData &mapdata,
+                             const QString &filename,
+                             QFile *const file,
+                             QObject *parent)
+    : AbstractMapStorage(mapdata, filename, file, parent)
+{}
+
+XmlMapStorage::~XmlMapStorage() = default;
+
+void XmlMapStorage::newData()
+{
+    qWarning() << "XmlMapStorage does not implement newData()";
+}
+
+bool XmlMapStorage::loadData()
+{
+    return false;
+}
+
+bool XmlMapStorage::mergeData()
+{
+    return false;
+}
+
+bool XmlMapStorage::saveData(bool baseMapOnly)
+{
+    log("Writing data to file ...");
+
+    // Collect the room and marker lists. The room list can't be acquired
+    // directly apparently and we have to go through a RoomSaver which receives
+    // them from a sort of callback function.
+    // The RoomSaver acts as a lock on the rooms.
+    ConstRoomList roomList;
+    roomList.reserve(m_mapData.getRoomsCount());
+
+    RoomSaver saver(m_mapData, roomList);
+    for (uint i = 0, n = m_mapData.getRoomsCount(); i < n; ++i) {
```

Comment:
Why not just cache `n` as `roomsCount` higher up in the code? You're already calling it several times and there is no need to "optimize" here in the loop really.

Reply:
makes sense

Reply:
done